core/dnsserver/server-grpc.go

```diff
@@ -24,9 +24,34 @@ type ServergRPC struct {
 	listenAddr net.Addr
 }
 
+type listenerTLS struct {
+	net.Listener
+	innerListener net.Listener
+	config        *tls.Config
+}
+
+// Dup implemenents caddy.Duppablelistener interface
+func (l listenerTLS) Dup() (net.Listener, error) {
+	file, err := l.innerListener.(*net.TCPListener).File()
```

Comment:
this is safe wrt to data races? Otherwise LGTM. Without a test would be good to update the PR with a comment showing it works.

thanks!

Reply:
@miekg , thanks for the review!

As regard to your question - this is very much the same code in the same code path as before, except that the implementation is now in CoreDNS instead of Caddy. The only difference is that the listener above is an underlying listener for the tls.Listener. Hope this allay your concern.

Totally agree the need to demonstrate that it works. I will include output of tests that I performed. Will add that to the PR soon.